### PR TITLE
Moving methods in cross_sections.F90 to attributes of Nuclide and Material objects

### PR DESCRIPTION
**Pull request by nelsonag**
To create a more object-based structure, this PR moves the cross section generating functions from the unbound cross_sections.F90 module to be bound objects of the Nuclide and Material objects.  This structure mimics that used for the equivalent data for the multi-group mode.  All tests pass and there seems to be no measurable hit to runtime on the problems I'm able to run on my machine.

### CODE REVIEW

**Comment by nelsonag** [comment]
You know what, no need to merge just yet.  I should also move some routines from physics.F90 in to Nuclide and Material like how sample_scatter is a bound method of Mgxs.

**Comment by nelsonag** [comment]
After doing some of the moving of methods from physics.F90 to type-bound methods of Material, Nuclide, and SAlphaBeta, I decided to abandon those efforts.  The code was no simpler, physics.F90 would still be pretty sizable, and it became confusing for why some methods were type bound while others werent since not all could easily be made type-bound.  So this PR is back to its original intent.

As a bonus I also included a change to physics.F90 that I came across during my attempt to move methods.  In the scatter routine, it would pass through every non-elastic reaction and ask 'are you a scatter?' every time.  I moved this to an initialization type of operation so this check is no longer performed.  On a BWR assembly problem with ~25 isotopes this resulted in a 5% speedup over the develop branch at the cost of each Nuclide owning an integer array as long as the number of inelastic scattering reactions that it contains.

**Comment by paulromano** [comment]
Thanks for the PR @nelsonag. I'll try to get to reviewing this soon. The code looks fine -- I just want to run some performance tests of my own to make sure there are no problems.

File: src/physics.F90 [hunk 1]
```diff
@@ -392,24 +391,14 @@ subroutine scatter(p, i_nuclide, i_nuc_mat)
...
-          associate (xs => rx % xs(i_temp))
+        associate (rx => nuc % reactions(i), &
+                   xs => nuc % reactions(i) % xs(i_temp))
             ! if energy is below threshold for this reaction, skip it
             if (i_grid < xs % threshold) cycle
 
             ! add to cumulative probability
             prob = prob + ((ONE - f)*xs % value(i_grid - xs % threshold + 1) &
                  + f*(xs % value(i_grid - xs % threshold + 2)))
-          end associate
```
Comment (smharper) [inline src/physics.F90]: Just one small request from me:  de-indent this block of code by two spaces.

**Comment by nelsonag** [comment]
Thanks! Done

File: src/nuclide_header.F90 [hunk 2]
```diff
@@ -291,6 +303,7 @@ subroutine nuclide_from_hdf5(this, group_id, temperature, method, tolerance, &
     real(8) :: temp_actual
     type(VectorInt) :: MTs
     type(VectorInt) :: temps_to_read
+    type(VectorInt) :: index_inelastic_scatter_vector
```
[the remark this answers is not in the document]
Comment (nelsonag) [reply src/nuclide_header.F90]: - [x]

File: src/nuclide_header.F90 [hunk 3]
```diff
@@ -458,10 +471,26 @@ subroutine nuclide_from_hdf5(this, group_id, temperature, method, tolerance, &
...
+      end if
+
       call close_group(rx_group)
     end do
     call close_group(rxs_group)
 
+    ! Recast to a regular array to save space
+    allocate(this % index_inelastic_scatter( &
+         index_inelastic_scatter_vector % size()))
+    this % index_inelastic_scatter = index_inelastic_scatter_vector % data(:)
```
[the remark this answers is not in the document]
Comment (nelsonag) [reply src/nuclide_header.F90]: Oops, good catch.

- [x]

File: src/nuclide_header.F90 [hunk 4]
```diff
@@ -458,10 +471,26 @@ subroutine nuclide_from_hdf5(this, group_id, temperature, method, tolerance, &
...
+    call index_inelastic_scatter_vector % clear()
```
[the remark this answers is not in the document]
Comment (nelsonag) [reply src/nuclide_header.F90]: - [x]

**Comment by paulromano** [comment]
Thanks again!

**Pull request merged**